File: app/src/main/java/nl/tudelft/cs4160/trustchain_android/KeyActivity.java
package nl.tudelft.cs4160.trustchain_android;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Base64;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import java.security.KeyPair;
import java.security.Security;

import nl.tudelft.cs4160.trustchain_android.Util.Key;

public class KeyActivity extends AppCompatActivity {

    static {
        Security.insertProviderAt(new org.spongycastle.jce.provider.BouncyCastleProvider(), 1);}

    private Button buttonNewKey;
    private Button signData;
    private TextView textPrivateKey;
    private TextView signedData;
    private Button verifySignature;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_key);
        init();
    }

    private void init() {
        buttonNewKey = (Button) findViewById(R.id.new_key);
        textPrivateKey = (TextView) findViewById(R.id.private_key);
        signData = (Button) findViewById(R.id.sign_data);
        signedData = (TextView ) findViewById(R.id.signed_data);
        verifySignature = (Button) findViewById(R.id.verify_sig);

        KeyPair kp = Key.loadKeys(getApplicationContext());
        if(kp == null) {
            kp = Key.createNewKeyPair();
            Key.saveKey(getApplicationContext(), Key.DEFAULT_PUB_KEY_FILE, kp.getPublic());
            Key.saveKey(getApplicationContext(), Key.DEFAULT_PRIV_KEY_FILE, kp.getPrivate());
        }
        textPrivateKey.setText(Base64.encodeToString(kp.getPrivate().getEncoded(), Base64.DEFAULT));

        verifySignature.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                KeyPair kp = Key.loadKeys(getApplicationContext());
                byte[] sig = Base64.decode(signedData.getText().toString(), Base64.DEFAULT);
                byte[] data = new byte[] {0x30, 0x30, 0x30, 0x30,0x30, 0x30, 0x30, 0x30};
                if(Key.verify(kp.getPublic(), data, sig)) {
                    Toast.makeText(getApplicationContext(), R.string.valid_signature, Toast.LENGTH_SHORT).show();
                } else {
                    Toast.makeText(getApplicationContext(), R.string.invalid_signature, Toast.LENGTH_SHORT).show();
                }
            }
        });

        buttonNewKey.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                KeyPair kp = Key.createNewKeyPair();
                Key.saveKey(getApplicationContext(), Key.DEFAULT_PUB_KEY_FILE, kp.getPublic());
                Key.saveKey(getApplicationContext(), Key.DEFAULT_PRIV_KEY_FILE, kp.getPrivate());
                textPrivateKey.setText(Base64.encodeToString(kp.getPrivate().getEncoded(), Base64.DEFAULT));

            }
        });

        signData.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                KeyPair kp = Key.loadKeys(getApplicationContext());
                byte[] sig = Key.sign( kp.getPrivate(), new byte[] {0x30, 0x30, 0x30, 0x30,0x30, 0x30, 0x30, 0x30});
                if(sig == null) {
                    System.out.println("No sig received");
                }
                signedData.setText(Base64.encodeToString(sig, Base64.DEFAULT));

            }
        });

    }


}

File: app/src/main/java/nl/tudelft/cs4160/trustchain_android/KeyActivity.kt
package nl.tudelft.cs4160.trustchain_android

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Base64
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_key.*
import nl.tudelft.cs4160.trustchain_android.Util.Key
import java.security.Security

class KeyActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_key)
        init()
    }

    private fun init() {
        val kp = Key.ensureKeysExist(applicationContext)
        privateKey.text = Base64.encodeToString(kp.private.encoded, Base64.DEFAULT)


        verifySignature.setOnClickListener {
            val kp = Key.loadKeys(applicationContext)
            val sig = Base64.decode(signedData.text.toString(), Base64.DEFAULT)
            if (Key.verify(kp.public, sampleData, sig)) {
                Toast.makeText(applicationContext, R.string.valid_signature, Toast.LENGTH_SHORT).show()
            } else {
                Toast.makeText(applicationContext, R.string.invalid_signature, Toast.LENGTH_SHORT).show()
            }
        }

        newKey.setOnClickListener {
            val kp = Key.createAndSaveKeys(applicationContext)
            privateKey.text = Base64.encodeToString(kp.private.encoded, Base64.DEFAULT)
        }

        signData.setOnClickListener {
            val kp = Key.loadKeys(applicationContext)
            val sig = Key.sign(kp!!.private, sampleData)
            if (sig == null) {
                println("No sig received")
            }
            signedData.text = Base64.encodeToString(sig, Base64.DEFAULT)
        }

    }

    companion object {
        val sampleData = byteArrayOf(0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30)

        init {
            Security.insertProviderAt(org.spongycastle.jce.provider.BouncyCastleProvider(), 1)
        }
    }


}
